### PR TITLE
Add docs for optimization functions in Gem Builder

### PR DESCRIPTION
**Pull request by alexanderahn**
---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1208299531873422

### CODE REVIEW

**Comment by neontty** [comment]
I have some recommendations incoming...

File: docs/Spark/extensibility/gem-builder/gem-builder.md [hunk 1]
```diff
@@ -320,14 +323,15 @@ Depending on what kind of Gem is being created, either a `Dialog` or a `DatasetD
...
+- The **Dataset Dialog**: The Dialog for a [Source/Target](../../gems/source-target/source-target.md) Gem is a `DatasetDialog` object. You will need to have `source` and `target` methods defined.
 
 Column Selector: This is a special property that you should add if you want to select the columns from UI and then highlight the used columns using the `onChange` function.
 It is recommended to try out this dialogue code in Gem builder UI and see how each of these elements looks in UI.
 
 ### Validation
 
-The `validate` method performs validation checks so that in the case where there's any issue with any inputs provided for the user an Error can be displayed. In our example case if the Filter condition is empty. Similarly, you can add any validation on your properties.
+The `validate` method performs validation checks for issues with user-provided inputs and displays errors or warnings if
+any are detected. In our example case if the Filter condition is empty. Similarly, you can add any validation on your properties.
```
Comment (neontty) [inline docs/Spark/extensibility/gem-builder/gem-builder.md:334]: this text doesnt' make any sense: 
> In our example case if the Filter condition is empty. Similarly, you can add any validation on your properties.

and this is not a very helpful example of validation. let's please fix this before closing the PR. may need help from a DE to get a better example of validation

Comment (alexanderahn) [reply docs/Spark/extensibility/gem-builder/gem-builder.md:334]: Thank you for catching this. I've rewritten it slightly.

This example existed before this PR so it's out of the original change scope. Do you mind if we track your request here instead? https://app.asana.com/0/1203457878958265/1208449446000978 I've reached out to folks already to see who can provide a better example.

**Comment by neontty** [comment]
@alexanderahn sorry I did not respond in time; it's fine to track the out-of-scope comments elsewhere. Thanks for writing this up!

**Pull request merged**